Google Family Link not seeing Sandboxed Play Services in Secondary User

Opening Family Link displays welcome screens but concluded through to main interface results in the app displaying Google Play Services not available and two notifications in the shade stating it won't work without them. 

Logs below:
`08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: FATAL EXCEPTION: main
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: Process: com.google.android.apps.kids.familylink, PID: 12963
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: java.lang.RuntimeException: Unable to start receiver com.google.android.libraries.notifications.entrypoints.restart.RestartReceiver: java.lang.SecurityException: Failed to find provider com.google.android.gsf.gservices for user 10; expected to find a valid ContentProvider for this authority
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:4036)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.app.ActivityThread.access$1400(ActivityThread.java:238)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1925)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7666)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at com.android.internal.os.ExecInit.main(ExecInit.java:48)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:399)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: Caused by: java.lang.SecurityException: Failed to find provider com.google.android.gsf.gservices for user 10; expected to find a valid ContentProvider for this authority
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:2373)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:2357)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2340)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2282)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.content.IContentService$Stub$Proxy.registerContentObserver(IContentService.java:1229)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.content.ContentResolver.registerContentObserver(ContentResolver.java:2627)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.content.ContentResolver.registerContentObserver(ContentResolver.java:2615)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at ivp.<init>(PG:2)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at ivp.a(PG:2)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at iwe.c(PG:7)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at nza.c(PG:1)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at hzt.onReceive(PG:27)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:4027)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	... 10 more
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: Caused by: android.os.RemoteException: Remote stack trace:
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at com.android.server.content.ContentService.registerContentObserver(ContentService.java:355)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.content.IContentService$Stub.onTransact(IContentService.java:482)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.os.Binder.execTransactInternal(Binder.java:1154)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 	at android.os.Binder.execTransact(Binder.java:1123)
08-15 10:41:01.106 1010245 12963 12963 E AndroidRuntime: 
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: FATAL EXCEPTION: GoogleApiHandler
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: Process: com.google.android.apps.kids.familylink, PID: 12963
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: java.lang.RuntimeException: Bad file descriptor
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at android.os.BinderProxy.transactNative(Native Method)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at android.os.BinderProxy.transact(BinderProxy.java:550)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at android.app.INotificationManager$Stub$Proxy.getNotificationChannel(INotificationManager.java:3883)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at android.app.NotificationManager.getNotificationChannel(NotificationManager.java:852)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at gay.e(PG:28)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at gfb.i(PG:2)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at gex.j(PG:18)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at gex.b(PG:1)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at gex.e(PG:7)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at gex.f(PG:7)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at gfb.handleMessage(PG:96)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
08-15 10:41:01.106 1010245 12963 13000 E AndroidRuntime: 	at android.os.HandlerThread.run(HandlerThread.java:67)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: Error reporting crash
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: java.lang.RuntimeException: Bad file descriptor
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at android.os.BinderProxy.transactNative(Native Method)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at android.os.BinderProxy.transact(BinderProxy.java:550)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at android.app.IActivityManager$Stub$Proxy.handleApplicationCrash(IActivityManager.java:5208)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$KillApplicationHandler.uncaughtException(RuntimeInit.java:158)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at jeg.b(PG:34)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at jef.uncaughtException(Unknown Source:4)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at ipc.uncaughtException(PG:81)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1073)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
08-15 10:41:01.131 1010245 12963 13000 E AndroidRuntime: 	at java.lang.Thread.dispatchUncaughtException(Thread.java:2203)`

`08-15 16:37:06.037  1000  1421 12952 I ActivityTaskManager: START u11 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.google.android.apps.kids.familylink/.home.HomeActivity bnds=[440,1947][641,2195]} from uid 1110150
08-15 16:37:06.075  1000  1421  1493 D CompatibilityChangeReporter: Compat change id reported: 135634846; UID 1110252; state: DISABLED
08-15 16:37:06.076  1000  1421  1493 D CompatibilityChangeReporter: Compat change id reported: 135754954; UID 1110252; state: ENABLED
08-15 16:37:06.077  1000  1421  1539 D CompatibilityChangeReporter: Compat change id reported: 143937733; UID 1110252; state: ENABLED
08-15 16:37:06.096  1000  1421  1539 I ActivityManager: Start proc 8288:com.google.android.apps.kids.familylink/u11a252 for pre-top-activity {com.google.android.apps.kids.familylink/com.google.android.apps.kids.familylink.home.HomeActivity}
08-15 16:37:06.203  1000  1421 13202 I ActivityTaskManager: Activity reported stop, but no longer stopping: ActivityRecord{8afc6e9 u0 com.spotify.music/.MainActivity t2965}
08-15 16:37:06.208  1000  1421 13202 I ActivityTaskManager: Activity reported stop, but no longer stopping: ActivityRecord{509d7ce u0 com.spotify.music/com.spotify.nowplaying.installation.music.NowPlayingActivity t2965}
08-15 16:37:06.549 1110252  8288  8288 I PackageParser: Skipping target and overlay pair com.android.settings and /product/overlay/SettingsOverlayG025J.apk: overlay ignored due to required system property: ro.boot.hardware.sku with value: G025J
08-15 16:37:06.553 1110252  8288  8288 I PackageParser: Skipping target and overlay pair com.android.settings and /product/overlay/SettingsOverlayG025M.apk: overlay ignored due to required system property: ro.boot.hardware.sku with value: G025M
08-15 16:37:06.857 1110252  8288  8337 E ActivityThread: Failed to find provider info for com.google.android.gms.chimera
08-15 16:37:07.005  1000  1421 11486 I ActivityTaskManager: START u11 {cmp=com.google.android.apps.kids.familylink/.account.selector.ParentInteractiveSelectorActivity} from uid 1110252
08-15 16:37:07.031  1000  1421 11486 W ActivityTaskManager: Tried to set launchTime (0) < mLastActivityLaunchTime (8741134)
08-15 16:37:07.078  1000  1421 13202 I ActivityTaskManager: START u11 {flg=0x10000 cmp=com.google.android.apps.kids.familylink/.warmwelcome.AnimatedWarmWelcomeActivity} from uid 1110252
08-15 16:37:07.091  1000  1421 13202 W ActivityTaskManager: Tried to set launchTime (0) < mLastActivityLaunchTime (8741678)
08-15 16:37:07.122  1000  1421  1421 D NotificationService: 11|com.google.android.apps.kids.familylink|10436|null|1110252: granting content://settings/system/notification_sound
08-15 16:37:07.127  1000  1421  1421 I chatty  : uid=1000(system) Binder:1421_3 identical 5 lines
08-15 16:37:07.127  1000  1421  1421 D NotificationService: 11|com.google.android.apps.kids.familylink|10436|null|1110252: granting content://settings/system/notification_sound
08-15 16:37:07.130  1000  1421  1421 W NotificationHistory: Attempted to add notif for locked/gone/disabled user 11
08-15 16:37:07.134  1000  1421  1421 E NotificationService: Muting recently noisy 11|com.google.android.apps.kids.familylink|10436|null|1110252
08-15 16:37:07.144  1000  1421  1421 I chatty  : uid=1000(system) Binder:1421_3 identical 1 line
08-15 16:37:07.156  1000  1421  1421 E NotificationService: Muting recently noisy 11|com.google.android.apps.kids.familylink|10436|null|1110252
08-15 16:37:07.172  1000  1421 11486 D CompatibilityChangeReporter: Compat change id reported: 136274596; UID 1110252; state: ENABLED
08-15 16:37:07.177  1000  1421  1421 E NotificationService: Muting recently noisy 11|com.google.android.apps.kids.familylink|10436|null|1110252
08-15 16:37:07.192  1000  1421  1421 I chatty  : uid=1000(system) Binder:1421_3 identical 1 line
08-15 16:37:07.200  1000  1421  1421 E NotificationService: Muting recently noisy 11|com.google.android.apps.kids.familylink|10436|null|1110252
08-15 16:37:09.636  1000  1421  4176 E TaskPersister: File error accessing recents directory (directory doesn't exist?).
08-15 16:37:10.138  1000  1421  1421 W NotificationHistory: Attempted to add notif for locked/gone/disabled user 11
08-15 16:37:10.145  1000  1421  1421 E NotificationService: Muting recently noisy 11|com.google.android.apps.kids.familylink|10436|null|1110252
08-15 16:37:11.353  1000  1421 12373 W OverlayManager: service 'idmap' died
08-15 16:37:13.320  1000  1421 12043 I ActivityTaskManager: START u11 {act=android.intent.action.VIEW dat=market://details?id=com.google.android.gms&pcampaignid=gcore_212880000-e-com.google.android.apps.kids.familylink-542249 flg=0x80000 pkg=com.android.vending cmp=com.android.vending/com.google.android.finsky.activities.MarketDeepLinkHandlerActivity} from uid 1110252
08-15 16:37:13.348  1000  1421 12399 W ActivityTaskManager: Tried to set launchTime (0) < mLastActivityLaunchTime (8026684)
08-15 16:37:13.453  1000  1421 12399 I ActivityTaskManager: START u11 {act=android.intent.action.VIEW dat=http://market.android.com/... cmp=com.android.vending/com.google.android.finsky.activities.MainActivity (has extras)} from uid 1110251
08-15 16:37:13.967  1000  1421 12043 D CompatibilityChangeReporter: Compat change id reported: 136274596; UID 1110251; state: ENABLED
08-15 16:37:15.747  1000  1421  4176 E TaskPersister: File error accessing recents directory (directory doesn't exist?).
08-15 16:37:17.515  1000  1421 12043 D CompatibilityChangeReporter: Compat change id reported: 135634846; UID 1110251; state: DISABLED
08-15 16:37:17.515  1000  1421 12043 D CompatibilityChangeReporter: Compat change id reported: 135754954; UID 1110251; state: ENABLED
08-15 16:37:17.516  1000  1421  1539 D CompatibilityChangeReporter: Compat change id reported: 143937733; UID 1110251; state: ENABLED
08-15 16:37:17.539  1000  1421  1539 I ActivityManager: Start proc 8409:com.android.vending:download_service/u11a251 for service {com.android.vending/com.google.android.finsky.ipcservers.downloadservice.DownloadServiceGrpcServerAndroidService}
08-15 16:37:17.950  1000  1421 12043 D CompatibilityChangeReporter: Compat change id reported: 136274596; UID 1110251; state: ENABLED
08-15 16:37:18.006 1110251  8409  8409 I PackageParser: Skipping target and overlay pair com.android.settings and /product/overlay/SettingsOverlayG025J.apk: overlay ignored due to required system property: ro.boot.hardware.sku with value: G025J
08-15 16:37:18.016 1110251  8409  8409 I PackageParser: Skipping target and overlay pair com.android.settings and /product/overlay/SettingsOverlayG025M.apk: overlay ignored due to required system property: ro.boot.hardware.sku with value: G025M
08-15 16:37:18.164  1000  1421 12043 D CompatibilityChangeReporter: Compat change id reported: 136219221; UID 1110251; state: ENABLED
08-15 16:37:18.198  1000  1421 12043 D ConnectivityService: requestNetwork for uid/pid:1110251/8409 NetworkRequest [ TRACK_DEFAULT id=294, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED Uid: 1110251 AdministratorUids: [] RequestorUid: 1110251 RequestorPackageName: com.android.vending] ]
08-15 16:37:18.199  1000  1421  2540 D ConnectivityService: NetReassign [294 : null → 109]
08-15 16:37:18.338  1000  1421 12043 D ConnectivityService: requestNetwork for uid/pid:1110251/8409 NetworkRequest [ TRACK_DEFAULT id=295, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED Uid: 1110251 AdministratorUids: [] RequestorUid: 1110251 RequestorPackageName: com.android.vending] ]
08-15 16:37:18.339  1000  1421  2540 D ConnectivityService: NetReassign [295 : null → 109]
08-15 16:37:18.343  1000  1421  2540 D ConnectivityService: NetReassign [no changes]
08-15 16:37:18.484  1000  1421  1421 D NotificationService: 11|com.android.vending|-56862258|null|1110251: granting content://settings/system/notification_sound
08-15 16:37:18.485  1000  1421  1421 I chatty  : uid=1000(system) Binder:1421_3 identical 5 lines
08-15 16:37:18.486  1000  1421  1421 D NotificationService: 11|com.android.vending|-56862258|null|1110251: granting content://settings/system/notification_sound
08-15 16:37:18.487  1000  1421  1421 W NotificationHistory: Attempted to add notif for locked/gone/disabled user 11`

--- a/core/java/android/app/compat/gms/GmsCompat.java
+++ b/core/java/android/app/compat/gms/GmsCompat.java
@@ -128,13 +128,23 @@ public final class GmsCompat {
         isDynamiteClientEnabled = isChangeEnabled("GMS_UNPRIVILEGED_DYNAMITE_CLIENT", GMS_UNPRIVILEGED_DYNAMITE_CLIENT);
     }
 
+    private static boolean validateCerts(Signature[] signatures) {
+        for (Signature signature : signatures) {
+            if (signature.toCharsString().equals(GmsInfo.SIGNING_CERT)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
     /**
      * Check whether the given app is unprivileged and part of the Google Play Services family.
      *
      * @hide
      */
-    public static boolean isGmsApp(String packageName, Signature[] signatures, boolean isPrivileged,
-            String sharedUserId) {
+    public static boolean isGmsApp(String packageName, Signature[] signatures,
+            Signature[] signatures2, boolean isPrivileged, String sharedUserId) {
         // Privileged GMS doesn't need any compatibility changes
         if (isPrivileged) {
             return false;
@@ -153,11 +163,13 @@ public final class GmsCompat {
         // Validate signature to avoid affecting apps like microG and Gcam Services Provider.
         // This isn't actually necessary from a security perspective because GMS doesn't get any
         // special privileges, but it's a failsafe to avoid unintentional compatibility issues.
-        boolean validCert = false;
-        for (Signature signature : signatures) {
-            if (signature.toCharsString().equals(GmsInfo.SIGNING_CERT)) {
-                validCert = true;
-            }
+        boolean validCert = validateCerts(signatures);
+
+        // Try past signing certificates if necessary. We iterate through two separate arrays here
+        // instead of concatenating them beforehand because this method gets called for every
+        // package on the system at boot, and thus needs to be efficient.
+        if (!validCert && signatures2 != null) {
+            validCert = validateCerts(signatures2);
         }
 
         return validCert;
@@ -184,7 +196,7 @@ public final class GmsCompat {
         Signature[] signatures = pkg.signingInfo.hasMultipleSigners() ?
                 pkg.signingInfo.getApkContentsSigners() :
                 pkg.signingInfo.getSigningCertificateHistory();
-        return isGmsApp(app.packageName, signatures, app.isPrivilegedApp(), pkg.sharedUserId);
+        return isGmsApp(app.packageName, signatures, null, app.isPrivilegedApp(), pkg.sharedUserId);
     }
 
     private static boolean isGmsInstalled(ApplicationInfo relatedApp) {

--- a/services/core/java/com/android/server/pm/AppsFilter.java
+++ b/services/core/java/com/android/server/pm/AppsFilter.java
@@ -565,13 +565,16 @@ public class AppsFilter {
             mQueriesViaComponentRequireRecompute = true;
         }
 
+        final boolean isGmsApp = GmsCompat.isGmsApp(newPkg.getPackageName(),
+                newPkg.getSigningDetails().signatures,
+                newPkg.getSigningDetails().pastSigningCertificates,
+                newPkg.isPrivileged(),
+                newPkgSetting.sharedUser != null ? newPkgSetting.sharedUser.name : null);
         final boolean newIsForceQueryable =
                 mForceQueryable.contains(newPkgSetting.appId)
                         /* shared user that is already force queryable */
                         || newPkgSetting.forceQueryableOverride /* adb override */
-                        || GmsCompat.isGmsApp(newPkg.getPackageName(),
-                            newPkg.getSigningDetails().signatures, newPkg.isPrivileged(),
-                            newPkgSetting.sharedUser != null ? newPkgSetting.sharedUser.name : null)
+                        || isGmsApp
                         || (newPkgSetting.isSystem() && (mSystemAppsQueryable
                         || newPkg.isForceQueryable()
                         || ArrayUtils.contains(mForceQueryableByDevicePackageNames,

